core/block.c

```diff
@@ -0,0 +1,398 @@
+        }
+    }
+    free_block_data(removed);
+}
+
+#ifdef LWM2M_RAW_BLOCK1_REQUESTS
+static
+uint8_t prv_coap_raw_block_handler(lwm2m_block_data_t ** pBlockDataHead,
+                               block_data_identifier_t identifier,
+                               uint16_t mid,
+                               block_type_t blockType,
+                               uint8_t * buffer,
+                               size_t length,
+                               uint16_t blockSize,
+                               uint32_t blockNum,
+                               bool blockMore)
+{
+    lwm2m_block_data_t * blockData = find_block_data(*pBlockDataHead, identifier, blockType);
+    
+    // manage new block transfer
+    if (blockNum == 0)
+    {
+        if (blockData == NULL)
+        {
+            blockData = prv_block_insert(pBlockDataHead, identifier, blockType);
+        }
+        else if (blockData->mid == mid)
+        {
+            return COAP_IGNORE;
+        }
+    }
+    // manage already started block1 transfer
+    else
+    {
+        if (blockData == NULL)
+        {
+            // we never receive the first block or block is out of order
+            return COAP_408_REQ_ENTITY_INCOMPLETE;
+        }
+
+        if (blockNum <= blockData->blockNum){
+            // this is a retransmissiion, ignore
+            return COAP_IGNORE;
+        }
+    }
+
+    // is it too large?
+    if (length > REST_MAX_CHUNK_SIZE) {
+        return COAP_413_ENTITY_TOO_LARGE;
+    }
+
+    blockData->blockNum = blockNum;
+    blockData->mid = mid;
+
+    if (blockMore)
+    {
+        return COAP_231_CONTINUE;
+    }
+    else
+    {
+        prv_block_data_delete(pBlockDataHead, identifier, blockType);
+
+        return NO_ERROR;
+    }
+}
+#endif
+
+static
+uint8_t prv_coap_block_handler(lwm2m_block_data_t ** pBlockDataHead,
+                               block_data_identifier_t identifier,
+                               block_type_t blockType,
+                               uint8_t * buffer,
+                               size_t length,
+                               uint16_t blockSize,
+                               uint32_t blockNum,
+                               bool blockMore,
+                               uint8_t ** outputBuffer,
+                               size_t * outputLength)
+{
+    lwm2m_block_data_t * blockData = find_block_data(*pBlockDataHead, identifier, blockType);
+    
+    // manage new block transfer
+    if (blockNum == 0)
+    {
+        if (blockData == NULL)
+        {
+            blockData = prv_block_insert(pBlockDataHead, identifier, blockType);
+        }
+        else
+        {
+            // there is already existing block for this resource, clear buffer
+            lwm2m_free(blockData->blockBuffer);
+            blockData->blockBuffer = NULL;
+            blockData->blockBufferSize = 0;
+        }
+
+        blockData->blockBuffer = lwm2m_malloc(length);
```

Comment:
No check that malloc returned != NULL, hardfault would occur in memcpy()

Reply:
Fixed